src/features/PowerShellNotebooks.ts

```diff
@@ -91,13 +92,31 @@ class PowerShellNotebookContentProvider implements vscode.NotebookContentProvide
         let cellKind: vscode.CellKind | undefined;
         let insideBlockComment: boolean = false;
 
+        // This dictates whether the BlockComment cell was read in with content on the same
+        // line as the opening <#. This is so we can preserve the format of the backing file on save.
+        let openBlockCommentOnOwnLine: boolean = false;
+
         // Iterate through all lines in a document (aka ps1 file) and group the lines
         // into cells (markdown or code) that will be rendered in Notebook mode.
         // tslint:disable-next-line: prefer-for-of
         for (let i = 0; i < lines.length; i++) {
             // Handle block comments
             if (insideBlockComment) {
-                if (lines[i] === "#>") {
+                if (lines[i].endsWith("#>")) {
```

Comment:
What if the line is something like:

```powershell
<# Hello #> Get-Process
```

Or

```powershell
Get-Process <# Here's how we do this #>
```

Reply:
None of that is gonna work for reasons before this change.

Reply:
Do you mean this change doesn't intend to fix those cases because there's some other reason why those cases won't work?

Reply:
Right so the middle two "just work" because they're gonna show up in a code cell.

The first one is broken because currently I don't handle `<#` and `#>` being on the same line.

The last one is broken and I'm not really sure how to handle it while also not changing the file being saved.

Reply:
It looks like the first two are particularly pernicious because those lines pass the `endsWith` checks in one of the conditions.

Is there a reason we have to attack it as lines rather than like this:

```typescript
let index = -1;
const codeCells = [];
const commentCells = [];
while (true) {
    const commentStart = content.indexOf('<#', index + 1);
    if (commentStart === -1) {
        break;
    }

    let commentEnd = content.indexOf('#>, commentStart + 2); // '#>'.length
    if (commentEnd === -1) {
        commentEnd === content.length - 1;
    }

    codeCells.push(content.substring(index, commentStart));
    codeCells.push(content.substring(commentStart, commentEnd));
    index = commentEnd + 2;
}
```

Reply:
> It looks like the first two are particularly pernicious

no it's only the first one. The `#>` check is only if `insideBlockComment`

Reply:
the bigger problem is saving though, not reading in as cells.

Reply:
As when cells are modified and we need to save them back to the script?

Reply:
Yeah exactly. That's the hard part because now you have to keep track of when the comment stops and where the code begins...

and also this:

```pwsh
Get-ChildItem <# look at this sick comment
it's amazing
#> <# oh shit now what happens #> Get-Foo <# 
oh god not another
#>
```

Reply:
Something like:

```typescript
let content = '';
for (let i = 0; i < codeCells.length; i++) {
    content += codeCells[i];
    content += commentCells[i];
   // There are various edge cases where the cell counts are different
}
```

Reply:
> and also this:

In that one I imagine there's three code cells, the middle one being empty

Reply:
> Something like:

What is this snippet trying to show?

Reply:
> What is this snippet trying to show?

How the file saves. Just a matter of putting the cells back together into the script

Reply:
but you don't have that info, you only have _cells_. https://github.com/PowerShell/vscode-powershell/blob/master/src/features/PowerShellNotebooks.ts#L225-L249

Reply:
maybe in the metadata we can store if they end in a newline...

Reply:
Well if we don't assume cells are broken up by newline, we wouldn't need to insert them again, no?

The cell itself would contain the newline (i.e. the code cells would start with newlines). Perhaps that's an issue though...

Also I notice we currently mangle the newlines in the file -- it might be worth changing that for Windows users

Reply:
> Well if we don't assume cells are broken up by newline, we wouldn't need to insert them again, no?

I think that would make this:

```pwsh
<#
asdf
#>
gci
<#
foo
#>
```

have a cell in Notebook mode look like:

```pwsh
[blank newline]
gci
```

which is awkward.

Reply:
> Also I notice we currently mangle the newlines in the file -- it might be worth changing that for Windows users

that's fixed in this PR

Reply:
There's the other part of the equation:

```
# Foo
# Bar
# Baz
```

^ That should be a markdown cell.

Reply:
> which is awkward.

Basically it sounds like block comments don't actually align well with the cell concept because they don't separate code segments in a straightforward way. They can begin and end on the same line, and many users use them for things like documenting individual parameters.

Cells seem to intrinsically (and intuitively) demand the notion of starting on a new line, which is something a line comment but not a block comment guarantees.

But given that people will, and probably should, expect block comments to work to some extent in notebooks, I would imagine the following scenarios:

```powershell
Get-Thing

    <# Separator #>

Get-Thing2
```

```powershell
Get-Thing

<# junk
  # =================
  junk #>

Get-Thing2
```

Where the block comment is in some way surrounded by newlines, it's a cell separator. One question there is whether we ignore spaces before or after the comment delineator.

Otherwise, block comments appear in the code cell blocks.

Reply:
But certainly all the scenarios in the earlier comment would appear in the code block, and not create new cells

Reply:
Any reason we can't say "the start of the block comment needs to start the line, and the end of the block comment needs to end a line"?

Reply:
I think that's reasonable yeah. So it's as simple as the `startsWith` and `endsWith` tests, and the only edge case is where they're on the same line

Reply:
Yeah, and for now, I'd like to defer that until someone opens an issue. It increases complexity because now "on save" I need to add `<#` and `#>`.

Reply:
Honestly, it's going to come up

Reply:
Last commit has support for `<# foo #>`

Reply:
but we live by the constraint that:

> your block comment must start a line with `<#` and end a line with `#>`.